Version tests

StefMa opened the pull request:
Fixes #168 

Since we are using internal classes of gradle and want to deliver high quality software we have to make sure that the plugin works with at least gradle 4.0 (and up).

For that I've added some tests which tests all gradle version we support (4.0 - 4.5) (without bugfix releases).

Currently 4.5. and android is broken...
We have to fix that but not in this PR 👍

tasomaniac commented on core/src/functionalTest/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,134 @@
+
+    @After
+    void tearDown() throws Exception {
+        temporaryFolder.delete()
```
`TemporaryFolder` rule should do this for us, shouldn't it?

tasomaniac commented on core/src/functionalTest/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    /**
+     * Android don't support only gradle 4.1 and up with plugin 3.0
```
What about
> Android Gradle Plugin 3.x supports only Gradle 4.1 and up

tasomaniac commented on core/src/functionalTest/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,134 @@
+    void test_withGradle41_shouldSucceed() {
+        def result = runner.withGradleVersion("4.1").build()
+
+        Assertions.assertThat(result.task(":bintrayUpload").outcome).isEqualTo(TaskOutcome.SUCCESS)
```
Should we static import `assertThat`?

tasomaniac commented on core/src/functionalTest/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,134 @@
+
+    @Test
+    void test_withGradle42_shouldSucceed() {
+        def result = runner.withGradleVersion("4.2").build()
```
What about a `Parameterized` test with 2 dimensional parameters. 
First param is the gradle version and second is the expected outcome.

So when a new version of gradle is released we can just modify params instead of copy pasting tests.

StefMa replied:
I don't really get it 🤔 

You mean having one test which runs different gradle versions and checks the output? 🤔 
Don't know if I like it because the report would be that "the test is broken" and not "test gradle42 is broken"...

tasomaniac replied:
You will write and maintain one test but actually they will be run separately. You can also name them.
Have a look at Parameterized test from JUnit. Also used in our static analysis plugin tests

StefMa replied:
Ah ok, now I get it.
I didn't know the [Parameterized test](https://github.com/junit-team/junit4/wiki/parameterized-tests) from junit. Thanks for the hint 👍

tasomaniac commented on core/src/functionalTest/groovy/com/novoda/gradle/release/JavaDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,103 @@
+                    """
+        )
+
+        def fakeSourceCode = new File(temporaryFolder.root, "src/main/java/HelloWorld.java")
```
Same comment here. 
Also please have a look at [TestProjectRule](https://github.com/novoda/gradle-static-analysis-plugin/blob/master/plugin/src/test/groovy/com/novoda/test/TestProjectRule.groovy). It basically makes it easier to create such projects in tests.

tasomaniac commented:
retest this please

tasomaniac commented:
The script failed with. Any idea?

> ci/prBuilder.sh: line 9: cd: ci/samples/: No such file or directory

tasomaniac commented on core/src/functionalTest/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,134 @@
+                    versionName "0.0.1"
+                }    
+                
+                lintOptions {
```
what about disabling lint completely? this will speed up testing time. It could be disabled here or we can add `-x lint` to the test build as an argument to skip the task.

StefMa commented on core/functionalTest.gradle:
```diff
@@ -0,0 +1,21 @@
+// Stolen from
+// https://github.com/gradle-guides/testing-gradle-plugins/blob/master/samples/code/url-verifier-plugin/gradle/functional-test.gradle
+
+sourceSets {
```
I'm also not 100% happy with that solution.
But I've read that https://guides.gradle.org/testing-gradle-plugins/ and IMO everything makes sense to me (having a test, functionalTest and integrationTest sourceSet).
But yes, we can move it...

About the `DefaultGradleRunner`:
No, the custom sourceSet is not needed to use the GradleRunner (DefaultGradleRunner is a internal API and **should not be used!**). We also apply the `gradleTestKit()`, but it is a little bit hidden ;)
Since we are using the `java-gradle-plugin` it will be automatically applied for us.
See https://docs.gradle.org/3.3/userguide/javaGradle_plugin.html:
> It automatically adds the gradleTestKit() dependency to the test compile configuration and generates a plugin classpath manifest file consumed by a GradleRunner instance if found.

But we have "another" issue with the PR in general.
I will explain it later in a upcoming commit.. 👍

tasomaniac replied:
I see. Thanks for the clarification. I just was curious why we need this setup.

StefMa commented:
So - I've basically rewrite the whole test stuff 😅 
But now it is way better IMO 👍 

One funny part to mention:
In the previous test I've used
```gradle
            buildscript {
                repositories {
                    jcenter()
                    google()
                }
                dependencies {
                    classpath 'com.android.tools.build:gradle:3.0.0'
                    classpath 'com.novoda:bintray-release:0.8.0'
                }
            }
            apply plugin: "com.android.library"
            apply plugin: "com.novoda.bintray-release"
```
Which **doesn't test our code but the already published code!**.
According to the GradleTestKit we have to use the `plugins{}` block to inject the **current** code into our tests.

tasomaniac commented:
Great. 🎉 @StefMa that sounds right, our static-analysis plugin tests also uses `plugins {}` block without the version declared.

tasomaniac commented on core/src/test/groovy/com/novoda/gradle/release/rule/GradleScriptTemplates.groovy:
```diff
@@ -0,0 +1,83 @@
+            }
+            
+            plugins {
+                id 'com.novoda.bintray-release' apply false
```
Hmmm, you can as well use `plugins` to use `com.android.library`, can't you? If you can do that, things would be simpler.

StefMa replied:
No we can't do that because the plugin isn't available in the `gradlePluginsPortal()`....
Otherwise we have to adjust the settings.gradle...
Don't want to do this...

blundell commented on core/src/test/groovy/com/novoda/gradle/release/AndroidDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Test
+    void testDifferentGradleVersionsAndOutcome() {
```
what is the outcome you are testing? 

`testOurUseOfInternalAPisWorksWithDifferentGradleVersions` ?

StefMa replied:
I don't just test the internal APIs only. I test the whole project.
Your test name sounds like "If we don't use internals anymore we can remove the test". Which I don't suggest.

Since we are using a Parameterized test I don't want to add "works" here because not all version works. And works would be the wrong outcome. Maybe we expect that it doesn't work (like with Gradle 4.0).

So I'm happy with my current method name.

blundell replied:
the name I suggested was an example of being explicit with describing the expectations.

Right now "outcome" isn't clear what is expected. If you are struggling to name it descriptively you could alternatively add javadoc, for instance with what you wrote above.

side note: this is a problem i find with parameterized tests. They may be "cool" at "speeding up" writing tests - but long term for legacy code developer supporters - they are confusing.

tasomaniac replied:
What about this

`givenGradleVersion_WhenProjectBuild_ShouldHaveExpectedOutcome`

tasomaniac replied:
@blundell I suggested `Parameterized` test. The reason is definitely not being cool or speeding up. It is more about maintainability. 

This class already has a single purpose. That's why I don't think it will not be confusing. This just tests if the build has expected outcome. We are not expecting any additional tests here.

If you do X different tests for X different cases here, you will have lots of duplication in the test code and also in test names. There is also no guarantee that future developers will keep them consistent.

blundell replied:
"`ShouldHaveExpectedOutcome`" is the problem

_ExpectedOutcome_ needs defining

tasomaniac replied:
To be that is obvious. The params are not even primitives. `GradleVerionsParams` is introduced to encapsulate the `gradle versions` and `outcome`.

tasomaniac replied:
Please come up with a suggestion to make the test naming more clear.

blundell replied:
```
                // Gradle 4.0 is not support by the Android Gradle Plugin 3.x
 +                new GradleVerionsParams("4.0", true),
 +                new GradleVerionsParams("4.1", TaskOutcome.SUCCESS),
 +                new GradleVerionsParams("4.2", TaskOutcome.SUCCESS),
 +                new GradleVerionsParams("4.3", TaskOutcome.SUCCESS),
 +                new GradleVerionsParams("4.4", TaskOutcome.SUCCESS),
 +                // TODO: Failure on Gradle 4.5. is **not** expected. It's failing because of changes in
 +                // the UsageContext in our AndroidLibrary class
 +                new GradleVerionsParams("4.5", true),
```

true vs success is not obvious

`testWeOnlySupportSpecificGradleVersions`

StefMa replied:
> "ShouldHaveExpectedOutcome" is the problem

I'm with you that "Outcome" is not a good naming here. 
With testing names we should have something like `calcularAdd_2plus5_shouldReturn7`.
The seven here is the expected output.
We should have someting like ` assertTrue(result, 7)`.
Mapping this to our test it is equivalent. Because we check if the `TaskOutcome` have a specific value (which we define in the Parameter method).

I think the name "Outcome" is here the problem?! But since the Gradle class `TaskOutcome` sounds outcome it is totally fine IMO.

`testWeOnlySupportSpecificGradleVersions` 
is also not that clear. Because we test ALSO the TaskOutput.
If we don't check the output with the `assertThat` but only the `build()` or `buildFail()` from the GradleRunner I would agree with that name. But as we also check the TaskOutput the method name removes some information ... :)

> true vs success is not obvious

I'm with you. It is a little bit crazy. In a previous implementation I had only the three params constructor. But since the `TaskOutcome` don't care if we already fail with `buildAndFail()`.
I can try to improve it with a named parameter... http://docs.groovy-lang.org/latest/html/documentation/#_named_argument_constructor

tasomaniac replied:
In a parameterized test, the output already has the name of the parameters for each run. This PR  will make it say 
```
test Gradle version: GradleVerionsParams{ gradleVersion='4.4.1', expectedGradleBuildFailure=false, expectedTaskOutcome=SUCCESS }
```

Given that, the test name should be about that particular param, not generic.  Considering that `testWeOnlySupportSpecificGradleVersions` will not be suitable because it does not really reflect what the test does. 

To be honest, I don't think the test name is super important because the param names in the test run makes it super clear.

blundell replied:
so whats the next steps?

blundell commented on core/src/test/groovy/com/novoda/gradle/release/JavaDifferentGradleVersions.groovy:
```diff
@@ -0,0 +1,51 @@
+import static org.assertj.core.api.Assertions.assertThat
+
+@RunWith(Parameterized.class)
+class JavaDifferentGradleVersions {
```
you could combine this with the Android test class so that the tests are all in the same place - makes it muc easier to understand / maintain

mr-archano replied:
I'm not sure is a good idea @blundell, as we need to create different synthetic projects and test separate outcomes too.

blundell replied:
I was scrolling between the two test classes and couldn't see much difference. Is there big setup overhead? Seems like they have some connescence with change was my thinking

mr-archano replied:
to achieve what you suggest we need to treat the projects as another dimension, that would male the test harder to understand and maintain IMO. I'd be happy with some duplication here.

StefMa replied:
I'm 50% with you. The code is more or less copied code which could be merged.

On the other side, the bintray-release plugin provides two different features. One is the **creating** of `jars` / `aars`. The other is the publishing to bintray... (more or less).
Only because of that reason it make sense to me to have two different test classes which tests two different possible scenarios.

It also brings much more complexity into our tests which can be avoided if we have two different test classes...

StefMa commented:
Just changed the test method name to the one which was suggest by @tasomaniac .
I think it was the best proposal until now...

I've also changed the Params class so that we can use namend constructor params...
Looks better IMO...

StefMa commented:
To be honest I think we talk way to much about this PR 😂

blundell commented:
thanks for the explanation and the patient discussion 👍

StefMa commented:
Jenkins reported green but jenkins says also:
![bildschirmfoto 2018-02-06 um 16 57 55](https://user-images.githubusercontent.com/10229883/35869241-e4bc46c4-0b5e-11e8-8977-87af4abdd442.png)

Strange

StefMa commented:
Fixed it temporary.. but the main issue is that it doesn't fail...
Normally all of other test should fail.

Whatever. I will take care of it in another PR 👍

blundell commented:
> Tested it with a GradleException and it breaks the tests... Seems we have to use it instead of a IllegalException

ah good to know!

tasomaniac commented:
Also take a look at `TaskExecutionException`. This requires you to provide the task which then gives the user and gradle more information where the failure happened.
